experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -71,6 +83,187 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+    // Connect commit to the correct control signal and end traversal
+    currOp->setOperand(1, ctrlSignal);
+    found_commit = true;
+  } else if (isa<handshake::ConditionalBranchOp>(currOp)) {
+    // Replicate a branch in the control path and use new control signal
+    MLIRContext *ctx = &getContext();
+    OpBuilder builder(ctx);
+
+    builder.setInsertionPointAfterValue(ctrlSignal);
+
+    // The Speculating Branch will discard the control token if the token
+    // in the speculative path is non-speculative.
+    Operation *branchDisc = builder.create<handshake::SpeculatingBranchOp>(
+        currOp->getLoc(), currOp->getResult(0), // specTagOperand
+        currOp->getOperand(0)                   // dataOperand
+    );
+    visited.insert(branchDisc);
+
+    // Connect a conditional branch at the true result of branchDisc
+    Operation *branchCond = builder.create<handshake::ConditionalBranchOp>(
+        branchDisc->getLoc(),
+        branchDisc->getResult(0), // conditionOperand
+        ctrlSignal                // dataOperand
+    );
+    visited.insert(branchCond);
+
+    // Follow the two branch results with a different control signal
+    for (int i = 0; i <= 1; ++i) {
+      for (Operation *succOp : currOp->getResult(i).getUsers()) {
+        Value ctrl = branchCond->getResult(i);
+        bool routed = routeBranchControlTraversal(visited, ctrl, succOp);
+        found_commit = found_commit || routed;
+      }
+    }
+
+    if (!found_commit) {
+      // Remove unused branch signal
+      branchCond->erase();
+      branchDisc->erase();
+    }
+
+  } else {
+    // Continue Traversal
+    for (Value res : currOp->getResults()) {
+      for (Operation *succOp : res.getUsers()) {
+        bool routed = routeBranchControlTraversal(visited, ctrlSignal, succOp);
+        found_commit = found_commit || routed;
+      }
+    }
+  }
+  return found_commit;
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceCommits() {
+  // Place commits and connect to the Speculator Commit Control Signal
+  Value commitCtrl = specOp.getCommitCtrl();
+  if (failed(placeUnits<handshake::SpecCommitOp>(commitCtrl)))
+    return failure();
+
+  // Create visited set
+  llvm::DenseSet<Operation *> visited;
+  visited.insert(specOp.getDataOut().getDefiningOp());
+
+  // Start traversal at the speculator output
+  for (Operation *succOp : specOp.getDataOut().getUsers()) {
+    routeBranchControlTraversal(visited, commitCtrl, succOp);
+  }
+
+  return success();
+}
+
+static LogicalResult findControlBranch(ModuleOp modOp, unsigned bb,
+                                       ConditionalBranchOp &cbranchOp) {
+  handshake::FuncOp funcOp = *modOp.getOps<handshake::FuncOp>().begin();
+  assert(funcOp && "funcOp not found");
+  auto handshakeBlocks = getLogicBBs(funcOp);
+
+  handshake::ConditionalBranchOp controlBranch;
+
+  // Iterate all Ops in the BB where we want to have Speculation
+  for (auto blockOp : handshakeBlocks.blocks.lookup(bb)) {
```

Comment:
is there a way to just do a query of get all condition branch ops in a bb, rather than iterating? 
(Most of my stuff is AST processing, and they have prebuilt optimized subtree querying, it seems the kind of thing MLIR might have)

Reply:
Here is a potentially cleaner piece of code that should do the same thing.
```cpp
for (auto condBrOp : funcOp.getOps<handshake::ConditionalBranchOp>()) {
  if (auto brBB = getLogicBB(condBrOp); !brBB || brBB != bb)
    continue;

  for (Value result : condBrOp->getResults()) {
    for (Operation *user : result.getUsers()) {
      if (isBackedge(result, user))
        return condBrOp;
    }
  }
}
return nullptr;
```

Note that I have changed the return type to `handshake::ConditionalBranchOp` since `nullptr` can indicate a failure in this case. This is less verbose than returning a `LogicalResult` and having an argument passed by reference.

Reply:
Now using Lucas suggestion for this function. It is checked if the result is a nullptr to emit an error.